pkg/multishare_reconciler/reconciler.go

```diff
@@ -0,0 +1,873 @@
+			instanceShare[parentURI] = append(instanceShare[parentURI], share)
+		}
+	}
+
+	return managedInstances, managedShares, instanceShare
+}
+
+// listMultishareOps reports all running or error ops related to multishare instances and share resources. The op target is of the form "projects/<>/locations/<>/instances/<>" or "projects/<>/locations/<>/instances/<>/shares/<>"
+func (recon *multishareReconciler) listMultishareResourceOps(ctx context.Context) ([]*Op, error) {
+	ops, err := recon.cloud.File.ListOps(ctx, &file.ListFilter{Project: recon.cloud.Project, Location: "-"})
+	if err != nil {
+		return nil, err
+	}
+
+	var finalops []*Op
+	for _, op := range ops {
+		if op.Done && op.Error == nil {
+			continue
+		}
+
+		if op.Metadata == nil {
+			continue
+		}
+
+		var meta filev1beta1.OperationMetadata
+		if err := json.Unmarshal(op.Metadata, &meta); err != nil {
+			klog.Errorf("Failed to parse metadata for op %s", op.Name)
+			continue
+		}
+
+		var err error
+		if op.Error != nil {
+			err = status.Error(codes.Code(op.Error.Code), op.Error.Message)
+		}
+
+		if file.IsInstanceTarget(meta.Target) {
+			finalops = append(finalops, &Op{Id: op.Name, Target: meta.Target, Type: util.ConvertInstanceOpVerbToType(meta.Verb), Err: err})
+		} else if file.IsShareTarget(meta.Target) {
+			finalops = append(finalops, &Op{Id: op.Name, Target: meta.Target, Type: util.ConvertShareOpVerbToType(meta.Verb), Err: err})
+		}
+		// TODO: Add other resource types if needed, when we support snapshot/backups.
+	}
+	return finalops, nil
+}
+
+func instanceFitShare(instanceInfo *v1alpha1.InstanceInfo, shareInfo *v1alpha1.ShareInfo) bool {
+	// instance needs to be:
+	// 1. not up for delete 2.of the same storage class and 3. has less than max number of shares assigned already
+	if instanceInfo.DeletionTimestamp != nil ||
+		instanceInfo.Labels[driver.ParamMultishareInstanceScLabel] != shareInfo.Spec.InstancePoolTag {
+		return false
+	}
+
+	if instanceInfo.Status != nil && len(instanceInfo.Status.ShareNames) >= util.MaxSharesPerInstance {
+		return false
+	}
+
+	return true
+}
+
+// instanceEmpty returns true if instanceInfo.Status.ShareNames has zero entries
+func instanceEmpty(instanceInfo *v1alpha1.InstanceInfo) bool {
+	if instanceInfo.Status == nil || len(instanceInfo.Status.ShareNames) != 0 {
+		return false
+	}
+	return true
+}
+
+func maybeAddCleanupFinalizer(instanceInfoClone *v1alpha1.InstanceInfo) (*v1alpha1.InstanceInfo, bool) {
+	if instanceInfoClone.DeletionTimestamp != nil {
+		klog.Infof("instanceInfo %q has deletionTimestamp so must have already been processed")
+		return instanceInfoClone, false
+	}
+	if len(instanceInfoClone.Finalizers) != 0 {
+		klog.Errorf("instanceInfo %q should not have any finalizer when it's DeletionTimestamp is not set but got %v. Attempting to remove all of them", instanceInfoClone.Name, instanceInfoClone.Finalizers)
+	}
+	instanceInfoClone.Finalizers = []string{util.FilestoreResourceCleanupFinalizer}
```

Comment:
Please elaborate in the definition of FilestoreResourceCleanupFinalizer how this is intended to be used.

Reply:
done